### DISCUSSION TITLE
possible new features contributions if interested

### DISCUSSION THREAD

**Discussion by bm16ton** [blackmagic-debug/blackmagic#1194] [Ideas]
Im pretty sure i already know the answer but i brought i2c-star (usb to i2c with mainlined kernel driver) into my personal bmp source, also added usb to gpio with irq support and wrote a kernel driver for it. Its for the i2c-star tried to bring to bmp but ran outa endpoints on blackpillv2 to bring it into bmp. Will probly try my 407blackboard i believe it has 2 more endpoints then blackpillv2. i wouldnt even ask except i see sum canbus stuff and i can program stms via i2c using rogers app and any i2c host with a kernel driver. I also added adc support for the blackpill the regular way plus added sum bulkendpoints and another kernel driver to dump the adc readings in binary to a dev/char device. I realize probly not interested but i thought id offer. Full disclosure im just statrting to learn how to program and have forest gumped my way thru by looking at other code/commits so its most likely messy and bad but it all works. Thank you for your awesome work! ill try and clean up the files and folders and have it on github today just in case.

**Comment by esden** [comment]
Hi @bm16ton,

This all sounds pretty cool, but it is all thoroughly out of scope of the project. I wish I could say let's add everything and kitchen sink to BMD, it might sound cool but as you can imagine this would cause some issues. Besides managing maintenance load, (after you contribute the new feature there is a continuing never ending work to keep things working and not break, so the more we add the more work it is for the core devs) it is important for me that BMD stays focused on doing what it does as well as it can. If we start adding too many extra functions that go beyond JTAG/SWD programming and debugging, it will distract us, and result in us being a multitool that does everything and nothing very well. There are plenty of projects with the goal of being a multitool, BMP is not that.

That said, you are welcome to keep a fork of the project with the added features and maintain that yourself, and offer it as an option for the community. I would only ask that you then also rename your firmware to something different than Black Magic Debug, to prevent confusion. Or at least add a suffix to the name to make sure it is clear that it is not the main official firmware.

Last but not least, we are working towards providing a way to easily update and flash firmware onto the official Black Magic Probe (aka. native). This will allow programming of alternate firmwares that provide different sets of functions from the official Black Magic Debug firmware. For example a dedicated SPI flash programmer firmware. It is still in very early development, and it will be quite a while before we are ready to deploy it, but this will be available eventually. We could consider listing alternate, third party Black Magic Debug builds like yours. But keep in mind, this will not include low cost dev board Black Magic Debug hosts, only the official hardware, so if your target is not the official hardware you can ignore that.

I bet it is not the answer you hoped for, but I hope you understand.

Cheers,
Piotr

**Comment by bm16ton** [comment]
Thank you so much for tge response! And no worries i knew all that before i
posted but just like (sorta) being apart of your community! A simple eadily
ignorable suggestion. I maintain outa tree kernel drivers for ftdi mpsse
usb to spi and usb to i2c drivers plus similiar ones for sum mcu's. I
highly suggest if not making a full fledged spi host firmware then at least
one that uses all the kernel drivers for spi flash and the like. I simply
attach my chip clip to the flash and plug in my ft2232 and if the
filesystem is supported by linux it automounts like athumb drive, but even
better is all of the various wonderful tools with years sometimes decades
of development behind them, all of which work without needing to know or
account for the fact its spi flash. May sound  silly but put that side by
side with any dedicated spi programmar/programs out there and its
impossible that you wouldnt choose the kernel driver option everytime. Not
to mention new chips/protocols/userland apps constantly being added wo you
having to do anything. Plus you would be the first real project to do so.
Just a thought but even for personal use i highly recommend giving it a go.
Thank ypu sir for ypur time and kindness ill keep you no longer :)

**Comment by esden** [comment]
I appreciate your kind response. I do like the idea of utilizing OS capabilities when possible. If that can result in a smooth user experience then it is worth doing. This is why BMD went the route of implementing a built in GDB server, so that the user does not have to install and configure as much. We will keep your suggestion in mind going forward. :)
